lib/dry/system/component.rb

```diff
@@ -1,49 +1,78 @@
-require 'inflecto'
+require 'dry-equalizer'
+require 'dry/system/loader'
 
 module Dry
   module System
     class Component
-      attr_reader :loader
-      attr_reader :identifier, :path, :file
+      include Dry::Equalizer(:identifier, :path)
 
-      def initialize(loader, input)
-        @loader = loader
+      PATH_SEPARATOR = '/'.freeze
+      WORD_REGEX = /\w+/.freeze
 
-        @identifier = input.to_s.gsub(loader.path_separator, loader.namespace_separator)
+      attr_reader :identifier, :path, :file, :options, :loader
 
-        if loader.default_namespace
-          re = /^#{Regexp.escape(loader.default_namespace)}#{Regexp.escape(loader.namespace_separator)}/
-          @identifier = @identifier.sub(re, '')
-        end
+      def self.new(name, options)
+        ns, sep = options.values_at(:namespace, :separator).map(&:to_s)
 
-        @path = input.to_s.gsub(loader.namespace_separator, loader.path_separator)
-        @file = "#{path}.rb"
+        identifier = name.to_s.scan(WORD_REGEX).reject { |s| ns == s }.join(sep)
```

Comment:
> what about if you wanted to pass a namespace like foo.bar and turn `foo.bar.baz` into `baz/`

We shouldn't allow that, the ns should not contain the separator

> Wouldn't it also convert `foo.bar.foo` into `bar/` with a namespace of `foo`?

It would turn it into `bar` actually, which would break.

...that's what I meant by "its not as robust as it could be" :) I'll add proper validation later though